src/stan/variational/base_family.hpp

```diff
@@ -31,6 +31,8 @@ namespace stan {
       Eigen::VectorXd transform(const Eigen::VectorXd& eta) const;
       template <class BaseRNG>
       void sample(BaseRNG& rng, Eigen::VectorXd& eta) const;
+      template <class BaseRNG>
+      void sample_lq(BaseRNG& rng, Eigen::VectorXd& eta, double& log_q) const;
```

Comment:
This method can be designed a lot simpler:
```
template <class BaseRNG>
double sample_lq(BaseRNG& rng, Eigen::VectorXd& eta) const;
```

That way you can test the output. The current function has to check for side-effects instead of looking at the output.

Why is the function called `sample_lq()` instead of `sample_log_q()`?

Reply:
Required: add tests for `sample_lq()`. 

And... I just realized maybe the `eta` is also output? Or is it just `log_q`? This would be much clearer if you included documentation. If not, everyone else is just guessing at what's going on.

Reply:
Hi Daniel. Thanks for reviewing. I have added documentations.  I also changed sample_lq() into  sample_log_q().  Here I let `sample_log_q()` to both sample from the normal approximation and calculate the log density of the sample. Alternatively, I can draw samples first from `sample()` and compute `double sample_log_q()` afterwards, but I think it will be more computationally efficient to compute the density on the fly.

As for the unit test, my understanding is that what the current unit tests in `unit/variational` do is to check the final output of VI. The posterior draws are returned by eta from 'sample_log_q()', in this sense it is as testable as any other internal functions.